File: threedollar-api-user-service/src/main/java/com/depromeet/threedollar/api/user/config/WebConfig.java
package com.depromeet.threedollar.api.user.config;

import java.util.List;

import org.springframework.context.MessageSource;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.support.ReloadableResourceBundleMessageSource;
import org.springframework.format.FormatterRegistry;
import org.springframework.validation.Validator;
import org.springframework.validation.beanvalidation.LocalValidatorFactoryBean;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.servlet.config.annotation.InterceptorRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

import com.depromeet.threedollar.api.core.config.converter.DecodeIdConverter;
import com.depromeet.threedollar.api.core.config.resolver.DeviceLocationArgumentResolver;
import com.depromeet.threedollar.api.core.config.resolver.MapLocationArgumentResolver;
import com.depromeet.threedollar.api.user.config.interceptor.AuthInterceptor;
import com.depromeet.threedollar.api.user.config.interceptor.UserMetadataInterceptor;
import com.depromeet.threedollar.api.user.config.resolver.UserIdResolver;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Configuration
public class WebConfig implements WebMvcConfigurer {

    private final AuthInterceptor authInterceptor;
    private final UserMetadataInterceptor userMetaDataInterceptor;

    private final UserIdResolver userIdResolver;
    private final DeviceLocationArgumentResolver deviceLocationArgumentResolver;
    private final MapLocationArgumentResolver mapLocationArgumentResolver;

    private final DecodeIdConverter decodeIdConverter;

    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        registry.addInterceptor(authInterceptor);
        registry.addInterceptor(userMetaDataInterceptor);
    }

    @Override
    public void addArgumentResolvers(List<HandlerMethodArgumentResolver> resolvers) {
        resolvers.addAll(List.of(userIdResolver, deviceLocationArgumentResolver, mapLocationArgumentResolver));
    }

    @Override
    public Validator getValidator() {
        LocalValidatorFactoryBean validatorFactoryBean = new LocalValidatorFactoryBean();
        validatorFactoryBean.setValidationMessageSource(validationMessageSource());
        return validatorFactoryBean;
    }

    @Bean
    public MessageSource validationMessageSource() {
        ReloadableResourceBundleMessageSource messageSource = new ReloadableResourceBundleMessageSource();
        messageSource.setBasename("classpath:/messages/validation");
        messageSource.setDefaultEncoding("UTF-8");
        messageSource.setCacheSeconds(60);
        return messageSource;
    }

    @Override
    public void addFormatters(FormatterRegistry registry) {
        registry.addConverter(decodeIdConverter);
    }

}

File: threedollar-api-admin-service/src/main/kotlin/com/depromeet/threedollar/api/adminservice/config/WebConfig.kt
package com.depromeet.threedollar.api.adminservice.config

import com.depromeet.threedollar.api.adminservice.config.interceptor.AuthInterceptor
import com.depromeet.threedollar.api.adminservice.config.resolver.AccountIdResolver
import com.depromeet.threedollar.api.core.common.config.converter.DecodeIdConverter
import org.springframework.context.MessageSource
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.support.ReloadableResourceBundleMessageSource
import org.springframework.format.FormatterRegistry
import org.springframework.validation.Validator
import org.springframework.validation.beanvalidation.LocalValidatorFactoryBean
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.servlet.config.annotation.CorsRegistry
import org.springframework.web.servlet.config.annotation.InterceptorRegistry
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer

@Configuration
class WebConfig(
    private val authInterceptor: AuthInterceptor,
    private val adminIdResolver: AccountIdResolver,
    private val decodeIdConverter: DecodeIdConverter,
) : WebMvcConfigurer {

    override fun addInterceptors(registry: InterceptorRegistry) {
        registry.addInterceptor(authInterceptor)
    }

    override fun addArgumentResolvers(resolvers: MutableList<HandlerMethodArgumentResolver>) {
        resolvers.add(adminIdResolver)
    }

    override fun addCorsMappings(registry: CorsRegistry) {
        registry.addMapping("/**")
            .allowedOrigins("https://admin.threedollars.co.kr", "https://admin.dev.threedollars.co.kr", "http://localhost:3000")
            .allowedMethods("GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "HEAD")
            .allowedHeaders("*")
            .allowCredentials(true)
    }

    override fun getValidator(): Validator {
        val validatorFactoryBean = LocalValidatorFactoryBean()
        validatorFactoryBean.setValidationMessageSource(validationMessageSource())
        return validatorFactoryBean
    }

    @Bean
    fun validationMessageSource(): MessageSource {
        val messageSource = ReloadableResourceBundleMessageSource()
        messageSource.setBasename("classpath:/messages/validation")
        messageSource.setDefaultEncoding("UTF-8")
        messageSource.setCacheSeconds(60)
        return messageSource
    }

    override fun addFormatters(registry: FormatterRegistry) {
        registry.addConverter(decodeIdConverter)
    }

}
